### ISSUE TITLE
Defining Residual Convolution Structures in Block from gluon Module

### ISSUE THREAD

**Issue by SumNeuron** [apache/incubator-mxnet#7946]
Please see this [S.O. post](https://stackoverflow.com/questions/46306782/mxnet-python3-defining-a-residual-convolution-structures-as-block-from-gluon-m) and if possible answer all questions / subquestions. If you prefer, the content can be copy-pasted here.

**Comment by kevinthesun** [comment]
1. self.ramp(self.conv(x)) vs mx.gluon.nn.Conv1D(activation='relu')(x) Yes. The latter applies a relu activation to the output of Conv1D.

2. mx.gluon.nn.Sequential is for grouping multiple layers into a block. Usually you don't need to explicitly define each layer as a class attribute. You can create a list to store all the layers you want to group and use a for loop to add all list elements into mx.gluon.nn.Sequential object.

3. Yes. Call forward on mx.gluon.nn.Sequential is equal to call forward on all child blocks, with topological order of computation graph.

**Comment by szha** [comment]
https://github.com/apache/incubator-mxnet/blob/master/python/mxnet/gluon/model_zoo/vision/resnet.py#L24-L25

**Comment by SumNeuron** [comment]
So to clarify:

```
net = nn.Sequential()
net.add(nn.Conv1D())
net.add(nn.Activation(activation="relu"))
net(x)

```

will automatically apply the activation layer to the preceding layer? e.g.

```
nn.Activation(activation="relu")(nn.Conv1D()(x))
```


Somewhat @szha ... it is a bit more complex than that.

Can I email you a file to look at ?

**Comment by szha** [comment]
I'd recommend pasting the simplest version here so that others who have similar questions can benefit from our conversation here.

**Comment by kevinthesun** [comment]
@SumNeuron  Another example of implementing skip connection with gluon is the U-net structure of generator in pixel2pixel GAN: https://github.com/zackchase/mxnet-the-straight-dope/blob/master/chapter14_generative-adversarial-networks/pixel2pixel.ipynb

**Comment by SumNeuron** [comment]
@szha normally I would. In this case I rather not.

Also, I am having trouble getting Conv1D to accept input.

Suppose, for simplicity,  I have a binary image:

```
b_img = [
[1,0,0,1,1,0,0,0],
[0,1,0,0,0,1,0,1],
[1,0,0,0,0,0,1,1]
]
```

and I want to run a kernel of 5 over  each channel

```
# what the kernel sees for each channel
channel1 --> [1,0,0,1,1], [0,0,1,1,0], [0,1,1,0,0], [1,1,0,0,0]
channel2 --> [0,1,0,0,0], [1,0,0,0,1], [0,0,0,1,0], [0,0,1,0,1]
channel3 --> [1,0,0,0,0], [0,0,0,0,0], [0,0,0,0,1], [0,0,0,1,1]
```

now if I feed a _single_ image into this layer, I get an error.
If I send a list of images then I don't.

**Comment by SumNeuron** [comment]
Why do I have to pass a batch into Conv1D?

Suppose I have a simple img:
```
b_img=[
   [0,1,0,1], # channel 1
   [1,0,1,1], # channel 2
   [0,1,0,0] # channel 3 (like RGB kind of but binary...)
]
```
where I want to run 1D convolutions on each channel (e.g. with a kernel of width 3 it would "see" each of the following per channel:
```
channel1 ---> [0,1,0] and [1,0,1]
channel2 ---> [1,0,1] and [0,1,1]
channel3 ---> [0,1,0] and [1,0,0]
```
so that I would end up with a `num_channels * kernel` sized output, e.g. 
here  it would be something like

```
[
[1,1,1],
[1,1,1],
[1,1,0]
]

```

throwing a single image of this sort into Conv1D throws errors, throwing multiple images does not...

**Comment by kevinthesun** [comment]
You will need bath size axis for it to work. A single image can be of shape (1, 3, 32, 32).

**Comment by SumNeuron** [comment]
@szha 

Here is the simplest example I was able to think of...

```
# Import statements
import mxnet as mx
import numpy as np
import math
import random
gpu_device=mx.gpu()
ctx = gpu_device
# ctx = mx.cpu()

# Lets get some concrete examples
# each channel is measuring some aspect of the signal
# for simplicity (and laziness), we will assume these
# are binary measurments (because it lines up nicely)

signal_1 = [ # contains class 1
    [0, 0, 0, 0, 0, 1], # channel 1
    [1, 0, 0, 0, 0, 1], # channel 2
    [0, 0, 1, 0, 0, 1]  # channel 3
]
signal_2 = [ # contains class 2
    [1, 0, 1, 0, 0, 0], # channel 1
    [1, 1, 0, 0, 0, 0], # channel 2
    [1, 0, 1, 0, 0, 1]  # channel 3
]
signal_3 = [ # contains nothing
    [1, 0, 0, 1, 0, 0], # channel 1
    [0, 0, 0, 1, 1, 1], # channel 2
    [1, 0, 1, 0, 0, 1]  # channel 3
]

# in this case, each channel's recordings is in it's own array
# with corresponding time values (t) increasing from left to right
# one could also restructure these signals such that time values
# are grouped together, such as
[   #1, 2, 3 channels
    [0, 1, 0], # t=0
    [0, 0, 0], # t=1
    [0, 0, 1], # t=2
    [0, 0, 0], # t=3
    [0, 0, 0], # t=4
    [1, 1, 1], # t=5
]
# which arguably is a boring colored line
# if one were to think of each combination of 
# channels as RGB. We will not do that!

# Going along this notion of thinking of this 
# signal as an image, the goal is to do
# do object detection within these signals.

# so if we go back to our dummy data signal 1
# ("image" 1), it might contain an object 
# (sub-signal / "dog") which starts at
# t=1 and ends at t=5
# signal_1 = [ # contains class 1 ("dog")
#
# time 0   1  2  3  4  5
#     [0, »0, 0, 0, 0, 1«], # channel 1
#     [1, »0, 0, 0, 0, 1«], # channel 2
#     [0, »0, 1, 0, 0, 1«]  # channel 3
# ]

# so to complete our dummy data lets
# add make up some location indices [start, stop]
signal_1_loc = [1,5]
signal_2_loc = [3,4]
signal_3_loc = [0,5] # this one contains nothing so it takes the whole signal


# drawing inspiration from the YOLO network
# We will split our "image" into two "cells"
# see slide 28 of https://docs.google.com/presentation/d/1kAa7NOamBt4calBU9iHgT8a86RRHz9Yz2oh4-GTdX6M/edit#slide=id.g15092aa245_0_230
# e.g. for signal_1
#   __cell 1___||__cell 2___
#    [0, 0, 0]-||-[0, 0, 1], # channel 1
#    [1, 0, 0]-||-[0, 0, 1], # channel 2
#    [0, 0, 1]-||-[0, 0, 1]  # channel 3


# unlike YOLO, "height" is irrelvant in regards to detection.
# because the "height" in our visualizaiton corresponds to the
# number of channels recording aspects of our signal
# so if we use our 3-channel signal as a image it has a
# height of 3 pixels.


# In YOLO's terminology, we divide our
# image into a grid (aka feature map)
# of cells

# Since we do not "care" about height
# our feature map can be 1 x 2 in this case
# e.g. as above where we split the signal
# into two cells, each taking up the entire "height"

# each "cell" in our grid should predict
# the probability of that cell being a type of class
# the x location of where the object starts
# the width of the object 
# (we drop y and height, because it doesnt matter here)

# we will do that for each box in each cell
# and then class probabilites afterwards 

feature_map = [1,2] # dimensions of the map ...
# (each value in this feature map corresponds to a cell) - refered to as grid in YOLO

boxes_per_cell = 2
predictions_per_box = 3 # P(object), x, width
classes = 3

# our output - per cell - has the dimensions
out_dim = feature_map + [boxes_per_cell * predictions_per_box + classes]
total_outputs = out_dim[0] *out_dim[1] * out_dim[2]
print("Output Dimension per: ", out_dim)
print("Total number of outputs: ", total_outputs)


# to visualize our output
vis_out =[
    [
        # Box1_P(object), Box1_x, Box1_width, Box2_P(object), Box2_x, Box2_width, P(class1 | obj), P(class2 | obj), P(class3 | obj)
        [0,0,0,0,0,0,0,0,0], # cell 1
        [0,0,0,0,0,0,0,0,0]  # cell 2
    ]
]
print(mx.nd.array(vis_out).shape)
mx.nd.array(vis_out)

# So let's make the "simplist" net possible 
# namely, a net that converts input
# into the desired output
net = mx.gluon.nn.Sequential()
net.add(mx.gluon.nn.Conv1D(channels=out_dim[2], kernel_size=3, strides=3, padding=0))
net.initialize()

# data needs to be 3d
data=mx.nd.array(signal_1).reshape((1,3,6))
print(net(data).shape)
net(data)
# so we get 1 x 9 x 2 instead of 1 x 2 x 9

# we will need some more "complete" input data
# e.g. for each signal we need P(object) for each cell
# we need the x and width
# we need P(class | object)
# thus
signal_1 = [ # contains class 1
    [0, 0, 0, 0, 0, 1], # channel 1
    [1, 0, 0, 0, 0, 1], # channel 2
    [0, 0, 1, 0, 0, 1]  # channel 3
]
# with
signal_1_loc = [1,5]

signal1 = {
    "cell_1": {
        # class 1, 2, 3
        "p(obj)": [1,0,0],
        "x": 1 #relative to box,
        "w": 2 # width inclusive
    },
    "cell_2": {
        # class 1, 2, 3
        "p(obj)": [1,0,0]
        "x": 0,
        "w": 3
    }
}

signal_2 = [ # contains class 2
    [1, 0, 1, 0, 0, 0], # channel 1
    [1, 1, 0, 0, 0, 0], # channel 2
    [1, 0, 1, 0, 0, 1]  # channel 3
]
#with 
signal_2_loc = [3,4]
signal2 = {
    "cell_1": {
        # class 1, 2, 3
        "p(obj)": [0,0,1]
        "x": 0,
        "w": 3
    },
    "cell_2": {
        # class 1, 2, 3
        "p(obj)": [0,1,0]
        "x": 1,
        "w": 2
    }
}

#and
signal_3 = [ # contains nothing
    [1, 0, 0, 1, 0, 0], # channel 1
    [0, 0, 0, 1, 1, 1], # channel 2
    [1, 0, 1, 0, 0, 1]  # channel 3
]
#width
signal_3_loc = [0,5]
signal3 = {
    "cell_1": {
        # class 1, 2, 3
        "p(obj)": [0,0,1],
        "x": 0,
        "w": 3
    },
    "cell_2": {
        # class 1, 2, 3
        "p(obj)": [0,0,1],
        "x": 0,
        "w": 3
    }
}


# and of course we need a custom backwards gradient for our last layer
# again reference https://docs.google.com/presentation/d/1kAa7NOamBt4calBU9iHgT8a86RRHz9Yz2oh4-GTdX6M/edit#slide=id.g15092aa245_0_230
# there are several steps in doing this gradient update
```

So @szha  how do I - in MXNet - pass this complex input  and implement the custom gradient :)

**Comment by szha** [comment]
Sorry for the late reply. You can probably use https://mxnet.incubator.apache.org/versions/master/api/python/autograd.html?highlight=autograd.function#mxnet.autograd.Function

**Comment by SumNeuron** [comment]
@szha  sorry for my late reply as well. I dont think that quite handles it...
do you have a native implementation of rCNN?

**Comment by szha** [comment]
There is an example of rcnn in the repo https://github.com/apache/incubator-mxnet/tree/master/example/rcnn

**Comment by ThomasDelteil** [comment]
@SumNeuron glancing quickly through the issue it seems that you want to format your data in the following way
`(N, C, T)` where `N` is the batch size, `C` the number of channels and `T` the length of signal time window (sometimes called `W` for width of your signal).
For example, for 3 channels and 64 measurements, using 32 convolutions kernels:
```python
N = 1
T = 64
C = 3
signal = mx.nd.ones((N, C, T))
conv1D = mx.gluon.nn.Conv1D(channels=32, kernel_size=3)
conv1D.initialize()
out = conv1D(signal)

out.shape
(1, 32, 62)
```

To answer your second question, have a look at GluonCV where you can find an implementation of Faster-RCNN and SSD. 
- [Faster-RCNN](https://gluon-cv.mxnet.io/build/examples_detection/train_ssd_voc.html#sphx-glr-build-examples-detection-train-ssd-voc-py)
- [SSD](https://gluon-cv.mxnet.io/build/examples_detection/train_ssd_voc.html)

@SumNeuron if you would like to follow up, please create a post on https://discuss.mxnet.io, thanks!
@szha could you please close the issue? Thanks!
